File: 034_find_first_and_last_position_of_element_in_sorted_array.java
class Solution {
    public int findLeft(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        // eventually right will be smaller than left to break the loop
        return left;
    }

    public int findRight(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] <= target) {  // <= for remove duplicate values
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        // eventually left will be larger than right to break the loop
        return right;
    }

    public int[] searchRange(int[] nums, int target) {
        int left = findLeft(nums, target);
        int right = findRight(nums, target);
        if (left <= right) {
            return new int[] {left, right};
        } else {
            return new int[] {-1, -1};
        }
    }
}

File: python/034_find_first_and_last_position_of_element_in_sorted_array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def findTarget(nums: List[int], target: int, isLeft: bool) -> bool:
            lo = 0
            hi = len(nums)
            while lo < hi:
                mid = (lo + hi) >> 1
                if nums[mid] > target or (isLeft and target == nums[mid]):
                    hi = mid
                else:
                    lo = mid + 1
            return lo

        left_idx = findTarget(nums, target, True)
        if left_idx == len(nums) or nums[left_idx] != target:
            return [-1, -1]
        return [left_idx, findTarget(nums, target, False) - 1]
